modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -261,106 +303,111 @@ module ChapelDynamicLoading {
+    // will be non-local.
     proc type create(path: string, out err: owned DynLoadError?) {
-      var ret: unmanaged chpl_BinaryInfo? = nil;
+      if checkForDynamicLoadingErrors(err) then return nil;
+
+      const store = chpl_binaryInfoStore.borrow();
+      var handleOnStoreLocale: c_ptr(void);
+
+      if const ret = store.checkForEntry(path, handleOnStoreLocale, err) {
+        // There was an existing entry, so return it.
+        return ret;
+      } else if err != nil {
+        // There was an error, so return nothing.
+        return nil;
+      }
 
-      if checkForDynamicLoadingErrors(err) then return ret;
+      var bin = new owned chpl_BinaryInfo(path, chpl_binaryInfoStore);
+      var errBuf = new chpl_localBuffer(owned DynLoadError?, numLocales);
+      var numErrs: atomic int;
 
-      on Locales[0] {
-        const store = chpl_binaryInfoStore.borrow();
-        var shouldCreateNewEntry = true;
+      // No entry and no error, so the handle used should not be 'nil'.
+      assert(handleOnStoreLocale != nil);
+      bin._systemHandles[store.locale.id] = handleOnStoreLocale;
 
-        // Start by opening a system handle on LOCALE-0.
-        var err0: owned DynLoadError?;
-        var ptr0 = localDynLoadOpen(path, err0);
+      if shouldFanOut {
+        const skip = store.locale;
+
+        coforall loc in fanToAll(skip=skip) with (ref errBuf) do on loc {
+          var err;
+          var ptr = localDynLoadOpen(path, err);
 
-        // TODO: Need to propagate/combine this error instead.
-        defer { if ptr0 then localDynLoadClose(ptr0, err0); }
+          // Ensure that the local pointer is closed (discarding the error).
+          defer { if ptr then localDynLoadClose(ptr, err); }
+
+          if ptr {
+            // Swap in pointer on 'bin'. This clears the variable.
+            const idx = here.id;
+            on bin do bin._systemHandles[idx] <=> ptr;
+            assert(ptr == nil);
+
+          } else if err == nil {
+            // TODO: Make it so all failure paths return an error.
+            const msg = 'Failed to open \'' + path + '\' on LOCALE-' +
+                        here.id:string;
+            err = new DynLoadError(msg);
+          }
 
-        if err0 != nil || ptr0 == nil {
-          if err0 != nil then err = err0;
-          shouldCreateNewEntry = false;
+          if err != nil {
+            const idx = here.id;
+            on errBuf do errBuf[idx] = err;
+            numErrs.add(1);
+          }
         }
+      }
 
-        // Check the LOCALE-0 store for an existing entry.
-        if shouldCreateNewEntry {
-          manage store.handleToInfo.read() as m {
-            shouldCreateNewEntry = !m.get(ptr0, ret);
+      if numErrs.read() > 0 {
+        // If there were errors, then report the first error.
+        for i in 0..<errBuf.size {
+          if errBuf[i] {
+            err = errBuf[i];
+            break;
           }
         }
```

Comment:
You can achieve the same affect by starting an atomic `erroredLocale` off as `-1` and then using `.compareAndSwap(-1, idx)`, and finally directly indexing into `errBuf` with `erroredLocale`. It would also save you a `for` loop.